File: backend/app/schemas/pdf.py
# backend/app/schemas/pdf.py
from pydantic import BaseModel, Field


class BlockRect(BaseModel):
    x0: float
    y0: float
    x1: float
    y1: float


class TextBlock(BaseModel):
    id: str
    page: int = Field(ge=0)
    rect: BlockRect
    text: str
    font: str
    size: float


class PageLayout(BaseModel):
    page: int = Field(ge=0)
    width: float
    height: float


class BlocksResponse(BaseModel):
    document_id: str
    page_count: int
    pages: list[PageLayout]
    blocks: list[TextBlock]


class UploadResponse(BaseModel):
    document_id: str
    filename: str


class ApplyChangeRequest(BaseModel):
    page: int = Field(ge=0)
    block_id: str
    rect: BlockRect
    text: str
    font: str | None = None
    font_size: float | None = Field(default=None, ge=4, le=72)
    bold: bool = False
    clear_old: bool = True


class DeleteBlockRequest(BaseModel):
    page: int = Field(ge=0)
    rect: BlockRect


class DeleteBlockResponse(BaseModel):
    document_id: str
    deleted: bool


class ApplyChangeResponse(BaseModel):
    document_id: str
    applied: bool
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/pdf.ts
// frontend/src/types/pdf.ts
export type BlockRect = {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
};

export type TextBlock = {
  id: string;
  page: number;
  rect: BlockRect;
  text: string;
  font: string;
  size: number;
};

export type PageLayout = {
  page: number;
  width: number;
  height: number;
};

export type UploadResponse = {
  document_id: string;
  filename: string;
};

export type BlocksResponse = {
  document_id: string;
  page_count: number;
  pages: PageLayout[];
  blocks: TextBlock[];
};

export type ApplyPayload = {
  page: number;
  block_id: string;
  rect: BlockRect;
  text: string;
  font: string;
  font_size: number;
  bold: boolean;
  clear_old: boolean;
};

export type ApplyResponse = {
  document_id: string;
  applied: boolean;
  warnings: string[];
};

export type DeletePayload = {
  page: number;
  rect: BlockRect;
};

export type DeleteResponse = {
  document_id: string;
  deleted: boolean;
};

export type EditorDraft = {
  blockId: string;
  page: number;
  rect: BlockRect;
  text: string;
  font: string;
  fontSize: number;
  bold: boolean;
};
